ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryCleanupService.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import io.confluent.ksql.schema.registry.SchemaRegistryUtil;
+import io.confluent.ksql.services.ServiceContext;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@code QueryCleanupService} helps cleanup external resources from queries
+ * out of the main line of query execution. This ensures that tasks that might
+ * take a long time don't happen on the CLI feedback path (such as cleaning up
+ * consumer groups).
+ *
+ * <p>NOTE: this cleanup service is intended to be used across threads and across
+ * real/sandboxed engines.</p>
+ */
+@SuppressWarnings("UnstableApiUsage")
+class QueryCleanupService extends AbstractExecutionThreadService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(QueryCleanupService.class);
+  private static final Runnable SHUTDOWN_SENTINEL = () -> { };
+
+  private final BlockingQueue<Runnable> cleanupTasks;
+
+  QueryCleanupService() {
+    cleanupTasks = new LinkedBlockingDeque<>();
+  }
+
+  @Override
+  protected void run() {
+    try {
+      while (true) {
+        final Runnable task = cleanupTasks.take();
+        if (task == SHUTDOWN_SENTINEL) {
+          return;
+        }
+
+        task.run();
+      }
+    } catch (final InterruptedException e) {
+      // gracefully exit if this method was interrupted and reset
+      // the interrupt flag
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    cleanupTasks.add(SHUTDOWN_SENTINEL);
+  }
+
+  public Set<String> pendingApplicationIds() {
+    return cleanupTasks.stream()
+        .filter(QueryCleanupTask.class::isInstance)
+        .map(QueryCleanupTask.class::cast)
+        .map(t -> t.appId).collect(ImmutableSet.toImmutableSet());
+  }
+
+  public void addCleanupTask(final QueryCleanupTask task) {
+    cleanupTasks.add(task);
+  }
+
+  @VisibleForTesting
+  void awaitAllPreviousProcessed() {
```

Comment:
If this is only used for testing, isn't it safer/better to move it to the testing code? You already have a method for adding tasks, all you need is a getter and then you can achieve the same, no?

Reply:
I tried this, but it's hard to do it without breaking the synchronization guarantees. Though maybe you gave me another idea I'll try out.